File: Code/Methods/Biomarkers/Core/NeuronBiomarkers.py
import os
import sys
import numpy as np
import pdb

# Biomarkers to manage and analyse neuronal simulation data and potentially experimental
# data too


def SplitTraceIntoAPs(t,v,threshold=0,timeThreshold=5):
	
    # Units for defaults
    # t, timeThreshold - ms
    # v, threshold - mV

    assert len(t) == len(v), "v and t length mismatch"

    crossings = []
    timeCrossings = np.array([])
    
    # Check this for loop ignores last element
    # Looks for crossings from below
    for i,voltage in enumerate(v[:-1]): # Check this works
        if voltage < threshold:
            if v[i+1] >= threshold:
                crossings.append(i)
                timeCrossings = np.append(timeCrossings,t[i])
                
    # For each crossing, remove all instances within 5 ms, leaving only the first crossing of the threshold 
    groupedCrossings = np.zeros(np.size(crossings),float) 
    
    for i in range(len(crossings)-1):
        if groupedCrossings[i] == 0:
            nearbyCrossings = np.array( (timeCrossings[i+1:] - timeCrossings[i]) < timeThreshold )
            # Assign 
            groupedCrossings[i+1:] += nearbyCrossings
            assert all(groupedCrossings < 2), "Grouped crossing grouped more than once"
            
    firstCrossIndices = np.where(groupedCrossings == 0)
    # Need to turn crossings into a numpy array to index it with np.where
    firstCrossings = np.array(crossings)[firstCrossIndices]
    numAPs = len(firstCrossings)
    assert numAPs >= 0, "Negative number of APs!"
    
    # Assign time and voltage to traces
    times = []
    voltages = []
    # If 1 or 0 APs, return 1 trace, otherwise...
    if (numAPs == 0) | (numAPs == 1):
        times.append(t)
        voltages.append(v)

    # If we have multiple APs, for each AP find the minimum value
    # of Vm before the next AP
    
    """ 
    There are some commented assumptions about where traces begin and end here. The core idea is that all data points in the trace have to be assigned to 1 and only 1 AP. If areas of quiescence are a problem for particular analysis methods, they will be stripped out by other specialised functions. 
    Our goal in this function is to divide up the trace without leaving any of it out, so that we have everything for any future analysis.
    """
    startIdx = np.zeros(numAPs,int)
    endIdx = np.zeros(numAPs,int)
    for AP in range(numAPs):
        if AP == 0:
            startIdx[0] = 0 # Start of first AP is beginning of trace
        else:
            startIdx[AP] = endIdx[AP-1]+1 # Start of all other APs is after last AP
    
        if AP == numAPs-1:
            endIdx[AP] = len(v)-1 # End of last AP is end of trace
        else:
            # Calculate end of this trace - end is minimum voltage of this trace
            # From threshold of this AP to just before beginning of next threshold
            voltageDuringCurrentAP = v[ firstCrossings[AP]:firstCrossings[AP+1] ]
            
            # Get min voltage index
            minVmIdx = np.argmin(voltageDuringCurrentAP)
            endIdx[AP] = firstCrossings[AP] + minVmIdx # Don't think I need to minus 1 because Python indices start at 0
            
        times.append(t[startIdx[AP]:endIdx[AP]+1])
        voltages.append(v[startIdx[AP]:endIdx[AP]+1]) # Add 1 to as Python slicing ends 1 before last index

    for i in range(len(startIdx)-1):
        assert endIdx[i]+1 == startIdx[i+1], "startIdx and endIdx don't match up."
        
    assert startIdx[0] == 0, "First AP doesn't start at beginning of trace."
    assert endIdx[-1] == len(v)-1, "Last AP doesn't end at end of trace."
        
    return{'times':times, 'voltages':voltages, 'startIndices':startIdx, 'endIndices':endIdx}  
        
    
def VoltageGradient(t,v):
    ### Is there a diff function in scipy or numpy?
    dVdt = np.zeros(len(v),float)
    for i in range(len(v)-1):
        dv = v[i+1]-v[i]
        dt = t[i+1]-t[i]
        dVdt[i] = dv/dt   
    return dVdt
        
    
# --- Biomarkers ---
def RMP(t,v):
    ### Is the a min function in scipy/numpy
    return min(v)
    
def StepRheobase(tracesFromStepTest,duration):

    return
    
def APPeak(t,v):
    ### Is there a max function
    return max(v)
    
def APRiseTime(t,v,threshold):
    dVdt = VoltageGradient(t,v)
    peak = APPeak(t,v)
    peakTime = peak[1]
    
    # If dVdt is a tuple, second part is gradient
    foundThreshold = []
    for i,gradient in enumerate(dVdt[1]): # Is dVdt a time vector as well?
        if gradient < threshold:
            if dVdt[1][i+1] > threshold:
                foundThreshold.append(i)
    
    numThresholds = len(foundThresholds)
    if numThresholds == 1:
        thresholdTime = t(foundThreshold[0])
        riseTime = peakTime - thresholdTime
        assert riseTime >=0, 'Rise time < 0!'
    elif numThresholds == 0:
        riseTime = 'N/A'
    elif numThresholds > 1:
        assert False, 'More than 1 threshold for rise time - APs may not be clearly separated.'
    return riseTime
        
def APSlopeMinMax(t,v):
    dVdt = VoltageGradient(t,v)
    slopeMin = min(dVdt)
    slopeMax = max(dVdt)
    ### Need mins and maxes
    return [slopeMin,slopeMax]
    
def APFullWidth(t,v,threshold):
    ups = []
    downs = []
    for i in range(len(v)-1):
        # Find ups (cross thresh from below)
        if v[i] < threshhold:
            if v[i+1] >= threshold: # Equals here so we trigger once if we flatten off at exactly threshold
                ups.append(i)
        #Find downs (cross thresh from above)
        if v[i] > threshold:
            if v[i+1] <= threshold:
                downs.append(i)
    numUps = len(ups)
    numDowns = len(downs)

    if (numUps < 1) | (numDowns < 1):
        # Not enough crossings
        fullWidth = 'N/A'
    elif (numUps == 1) & (numDowns == 1): 
        # One crossing of threshold each way
        fullWidth = t[downs[0]] - trace[ups[0]]
        
    elif (numUps > 1) | (numDowns > 1):
        # Too many crossings
        # Find earliest crossing from below
        # and latest crossing from above
        # to calculate full width
        earliestUp = ups[0]
        latestDown = downs[-1]
        fullWidth = t[latestDown] - t[earliestUp]
    
    return fullWidth

def FitAfterHyperpolarisation(t,v,t2,v2):

    maxIdx = []
    maxIdx[0] = np.argmax(v) # Get idx of max(v)
    maxIdx[1]= np.argmax(v2)# Get idx of max(v2)
    
    workingTime = np.concatenate((t[maxIdx[0]:],t2[:maxIdx[1]+1]),0) ### join t[maxIdx1:] up to t2[1:maxIdx[1]]
    workingVoltage = np.concatenate((v[maxIdx[0]:],v2[:maxIdx[1]+1]),0) ### join
       
    # AHP amplitude
    amp = min(workingVoltage)
    ampIdx = np.argmin(workingVoltage)
    
    # AHP time constant
    # TO DO!!
       
    return [amp,tau]
   
 

File: Code/Methods/Biomarkers/NeuronBiomarkers.m
% NeuronBiomarkers
% 09/12/15
% function [biomarkers biomarkerNames] = NeuronBiomarkers()
function NeuronBiomarkers()
% Collect all neuronal biomarkers we are using
% Common format: V is a 2 column vector, V(:,1) is time, V(:,2) is voltage.

% Necessary data for biomarker calculation
% Set of traces following protocol
% Stimulus levels for each trace

% Necessary processing for biomarker calculation
% Each trace checked for AP, if AP(s) exist, indices for each AP.

% Diagnostics:

% -- Load trace file and sort into individualrdy APs --
filename = 'basicSpikes.dat';
baseTrace =  OpenNeuronTrace(filename);

assert(size(baseTrace,2) == 2);
assert(size(baseTrace,1) > 2);

% TestSplitTraceIntoAPs(baseTrace);

traces = SplitTraceIntoAPs(baseTrace);


% Sort baseTrace into individual traces

% -- Run biomarkers on each trace and compare answers to my estimates --



% --- Analyse special APs ---

% Resting AP (restingAP)
% No stimulus, for finding resting potential

% Step test
% Series of APs with different stimulus current amplitudes to find AP
% threshold, rheobase etc. Store in a big matrix.

stepIncrement = 50; % pA
stepDuration = 800; % ms
numSteps = 80;

% Ramp test
% Series of APs with different ramp currents (from 0 to stimAmp over
% rampDuration)

rampIncrement = 50; % pA
rampDuration = 500; % ms
numSteps = 80;


disp('Go!')
end

% --|| SUBFUNCTIONS ||--
function trace = OpenNeuronTrace(filename)

trace = importdata(filename);

end

% Function to split a trace into its constituent APs
function [traces numAPs] = SplitTraceIntoAPs(trace)

% Find all APs (points where trace crosses voltage threshold, excluding points within the
% time threshold caused by experimental noise)

% ///MAGIC NUMBERS\\\
threshold = 0; % Manual threshold (mV)
timeThreshold = 5; % Threshold to group crossings of threshold (ms)

% Find each instance of voltage threshold crossing from below
crossings = [];
for i = 1:length(trace(:,1))-1
    
    if trace(i,2) < threshold;
        if trace(i+1,2) > threshold;
            crossings(end+1,1) = i;
        end
    end
end

% For each crossing, remove all instances within 5 ms, leaving only the
% first crossing of the threshold
timeCrossings = trace(crossings,1);
groupedCrossings = zeros(size(crossings));
for i = 1:length(crossings)-1
    
    % If not part of a group or the first instance of crossing the threshold
    % in a group of crossings, then continue to process, otherwise move on
    if groupedCrossings(i) == 0
        nearbyCrossings = (timeCrossings(i+1:end) - timeCrossings(i)) < timeThreshold; % Future crossings occurring in time threshold
        groupedCrossings(i+1:end) = groupedCrossings(i+1:end) + nearbyCrossings; % Assign them
        assert(all(groupedCrossings < 2)); % Make sure we are properly skipping all the crossings that are counted as groups
    end
    
end
firstCrossIndices = find(groupedCrossings == 0);

% Store only the first instance from each 5 ms interval starting at first point where voltage crossed
% the threshold from below without having previously crossed in last 5 ms
% (assuming time threshold = 5)
firstCrossings = crossings(firstCrossIndices);

% Compute number of APs
numAPs = length(firstCrossings);

% If num APs is 1 or 0, return one trace, the whole trace.
% Otherwise:
if numAPs == 0
    traces = trace;
    numAPs = 0;
    return
elseif numAPs == 1
    traces = trace;
    numAPs = 1;
    return
end

% For each AP, find the minimum value of Vm before the next AP.
assert(numAPs > 1);
for i = 1:numAPs
    
    if i == 1
        startIdx(i) = 1;
    else
        startIdx(i) = endIdx(i-1)+1;
    end
    
    if i == numAPs
        endIdx(i) = length(trace);
    else
        % Calculate end of this trace
        
        % Get voltage from this AP's crossing to next AP's crossing
        curAPUpIndex = firstCrossings(i);
        nextAPUpIndex = firstCrossings(i+1);
        voltageDuringCurrentAP = trace(curAPUpIndex:nextAPUpIndex,2);
        
        % Find the minimum voltage and use it to calculate end index
        [minVm minVmIdx] = min(voltageDuringCurrentAP);
        endIdx(i) = minVmIdx + curAPUpIndex-1; % Add back in previous indices
    end
    
    % Add the AP to the cell of APs
    traces{i} = trace(startIdx(i):endIdx(i),:);
    
end

assert(length(traces) == numAPs);
end

% --Gradient--
function dVdt = VoltageGradient(trace)
dt = diff(trace(:,1));
dV = diff(trace(:,2));
dVdt = [trace(1:end-1,1) dV./dt]; % If dt is in ms and dV in mV, units will be V/s
end

% --Biomarkers--

% Resting Potential
function restingVm = RMP(trace)
V = trace;
restinVm = min(V(:,2));
end

% Ramp AP
function APs = RampAPs()

end
% AP Threshold

% Step rheobase

function out = StepRheobase(stepTestTraces,duration)
% Processes a step test to find the first AP

% for i = 1:length(stepTest)
% if Find an AP
% id = i
% break
% end
%
% out = stepIncrement * i;

end

% AP Peak
function [peak peakTime] = APPeak(trace)

[peak peakIdx] = max(trace(:,2));
peakTime = trace(peakIdx,1);

end

% AP Rise time
% Time from dVdt above a certain threshold, up to when AP peak was reached
function riseTime = RiseTime(trace,threshold)

% Get dVdt
dVdt = VoltageGradient(trace)
% Find max
[peak peakTime] = APPeak(trace)

% Find threshold (if it occurs)
foundThreshold = [];
for i = 1:length(dVdt)-1
    if dVdt(i,2) < threshold
        if dVdt(i+1,2) > threshold
            foundThreshold(end+1) = i;
        end
    end
    
end

numThresholds = length(foundThreshold);
if numThreshold == 1
    % Find time at threshold
    thresholdTime = trace(foundThreshold(1),1);
    % Find diffrence from peak time
    riseTime = peakTime - thresholdTime;
    assert(riseTime >= 0);
elseif numThresholds == 0
    riseTime = 'N/A'
elseif numThresholds > 1
    error('More than 1 threshold for rise time - APs possibly not properly separated')
    
end

end

% AP slope max/min
function [slopeMin slopeMax] = APSlopeMinMax(trace)

dVdt = VoltageGradient(trace);
slopeMin = min(dVdt(:,2));
slopeMax = max(dVdt(:,2));

end

% AP width full
function [fullWidth numUps numDowns] = APFullWidth(trace,threshold)

% Measure width of AP in ms at threshold (default from Davidson et al. 2014
% is 5 mV.

% Find intersections at threshold. If less than 1 up and 1 down
% intersection return "N/A".
% Otherwise, return the difference between earliest up and latest down
% intersection.

ups = [];
downs = [];
for i = 1:length(trace)-1
    
    % Find the ups (crossing theshold from below)
    
    % Find the downs (crossing threshold from above)
    
end

% Count them
numUps = length(ups);
numDowns = length(downs);

% Calculate width
if numUps < 1 || numDowns < 1 % Not enough crossings
    fullWidth = 'N/A'
elseif numUps == 1 && numDowns == 1 % Exactly one crossing of threshold each way
    fullWidth = trace(downs(1),1) - trace(ups(1),1);
    
elseif numUps > 1 || numDowns > 1 % More than one threshold crossing one or both ways
    
    % Ups and downs will be found in chronological order, so find earliest
    % crossing from below, and latest crossing from above to calculate full
    % width.
    earliestUp = ups(1);
    latestDown = downs(end);
    fullWidth = trace(latestDown,1) - trace(earliestUp,1);
end

end

function [amp tau] = FitAfterHyperpolarisation(trace,nextTrace)

% Fit an afterhyperpolarisation curve to the end of an AP using the period
% between the peak of the current trace and the peak of the next trace.

% Reconstruct the repolarising period
maxIdx = [];
[~, maxIdx(1)] = max(trace(:,2));
[~, maxIdx(2)] = max(nextTrace(:,2))

workingTrace = [trace(maxIdx(1):end,:); nextTrace(1:maxIdx(2),:)];

% AHP amplitude
[amp ampI] = min(workingTrace);

% AHP time constant
% Divide up and fit somehow. Need to discard upstroke.



end
% |||---- TESTS ----|||

    function TestSplitTraceIntoAPs(trace)
        
        [traces numAPs] = SplitTraceIntoAPs(trace);
        numAPs
        
        fig
        hold on
        for i = 1:numAPs
            plot(traces{i}(:,1),traces{i}(:,2),'color',rand(3,1))
        end
        
        disp('TestSplitTraceIntoAPs complete')
    end